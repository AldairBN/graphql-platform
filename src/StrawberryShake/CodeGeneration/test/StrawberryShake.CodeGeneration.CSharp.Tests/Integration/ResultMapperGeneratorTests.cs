using System.Text;
using System.Threading.Tasks;
using Snapshooter.Xunit;
using StrawberryShake.CodeGeneration;
using StrawberryShake.CodeGeneration.CSharp;
using Xunit;

namespace StrawberryShake.Integration
{
    public class ResultMapperGeneratorTests
    {
        readonly StringBuilder _stringBuilder;
        readonly CodeWriter _codeWriter;
        readonly ResultFromEntityTypeMapperGenerator _generator;

        public ResultMapperGeneratorTests()
        {
            _stringBuilder = new StringBuilder();
            _codeWriter = new CodeWriter(_stringBuilder);
            _generator = new ResultFromEntityTypeMapperGenerator();
        }

        [Fact]
        public async Task GenerateResultMapper_Droid()
        {
            await _generator.WriteAsync(
                _codeWriter,
                IntegrationDescriptors.CreateDroidNamedTypeDescriptor());

            _stringBuilder.ToString().MatchSnapshot();
        }

        [Fact]
        public async Task GenerateResultMapper_DroidHero()
        {
            await _generator.WriteAsync(
                _codeWriter,
                IntegrationDescriptors.CreateDroidHeroNamedTypeDescriptor());

            _stringBuilder.ToString().MatchSnapshot();
        }


        [Fact]
        public async Task GenerateResultMapper_Human()
        {
            await _generator.WriteAsync(
                _codeWriter,
                IntegrationDescriptors.CreateHumanNamedTypeDescriptor());

            _stringBuilder.ToString().MatchSnapshot();
        }

        [Fact]
        public async Task GenerateResultMapper_HumanHero()
        {
            await _generator.WriteAsync(
                _codeWriter,
                IntegrationDescriptors.CreateHumanHeroNamedTypeDescriptor());

            _stringBuilder.ToString().MatchSnapshot();
        }

        [Fact]
        public async Task GenerateResultMapper_FriendsConnection()
        {
            await _generator.WriteAsync(
                _codeWriter,
                IntegrationDescriptors.CreateFriendsConnectionDescriptor());

            _stringBuilder.ToString().MatchSnapshot();
        }
    }
}
