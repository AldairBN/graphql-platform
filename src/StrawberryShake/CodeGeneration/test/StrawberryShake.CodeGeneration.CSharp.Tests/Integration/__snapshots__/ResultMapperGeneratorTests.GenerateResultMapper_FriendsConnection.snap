namespace StarWarsClient
{
    [global::System.CodeDom.Compiler.GeneratedCode("StrawberryShake", "11.0.0")]
    public partial class FriendsConnectionFromDataMapper
        : IEntityMapper<FriendsConnectionData, FriendsConnection>
    {
        private readonly IEntityStore _entityStore;
        private readonly IEntityMapper<DroidEntity, Droid> _droidFromDroidEntityMapper;
        private readonly IEntityMapper<HumanEntity, Human> _humanFromHumanEntityMapper;

        public FriendsConnectionFromDataMapper(
            IEntityStore entityStore,
            IEntityMapper<DroidEntity, Droid> droidFromDroidEntityMapper,
            IEntityMapper<HumanEntity, Human> humanFromHumanEntityMapper)
        {
            _entityStore = entityStore
                 ?? throw new ArgumentNullException(nameof(entityStore));
            _droidFromDroidEntityMapper = droidFromDroidEntityMapper
                 ?? throw new ArgumentNullException(nameof(droidFromDroidEntityMapper));
            _humanFromHumanEntityMapper = humanFromHumanEntityMapper
                 ?? throw new ArgumentNullException(nameof(humanFromHumanEntityMapper));
        }

        public FriendsConnection Map(FriendsConnectionData entity)
        {
            var nodes = new List<ICharacter>();
            foreach (var nodesItem in entity.Nodes)
            {
                ICharacter mappedItem = default!;

                if (nodesItem.Name.Equals("Droid", StringComparison.Ordinal))
                {
                    mappedItem = _droidFromDroidEntityMapper.Map(_entityStore.GetEntity<DroidEntity>(nodesItem));

                }
                else if (nodesItem.Name.Equals("Human", StringComparison.Ordinal))
                {
                    mappedItem = _humanFromHumanEntityMapper.Map(_entityStore.GetEntity<HumanEntity>(nodesItem));

                }
                else {
                    throw new NotSupportedException();
                }

                nodes.Add(mappedItem);
            }

            return new FriendsConnection(
                nodes,
                entity.TotalCount
            );
        }
    }
}
